lib/internal/modules/helpers.js

```diff
@@ -347,6 +348,21 @@ function parseTypeScript(source, options) {
   }
 }
 
+/**
+ *
+ * @param {string} filename
+ * @returns {boolean} Returns true if the nearest package.json is private.
+ */
+function isPrivateNodeModule(filename) {
+  const packageJsonReader = require('internal/modules/package_json_reader');
+  const resolved = StringPrototypeStartsWith(filename, 'file://') ? filename : pathToFileURL(filename).href;
+  const packageConfig = packageJsonReader.getPackageScopeConfig(resolved);
```

Comment:
should we cache already analized paths?
Like everything under `node_modules/foo` once we have already read read `node_modules/foo/package.json`